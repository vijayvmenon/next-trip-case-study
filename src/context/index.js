import { useContext } from 'react';
import AppContext, { Context } from './context';

export default AppContext;
export { Context };
export const useMyContext = () => useContext(Context);
