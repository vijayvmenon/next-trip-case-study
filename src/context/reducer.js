import constants from '../constants';
import ENTranslations from '../locales/en.json';
import { localStorageExists } from '../utils';

const localStorageLanguage =
  localStorageExists && localStorage.getItem('LANGUAGE');
const initialState = {
  tableExpanded: false,
  hideHeader: false,
  snackbar: { value: '', text: '' },
  language: localStorageLanguage || 'en',
  translations: localStorageLanguage
    ? constants.LANGUAGE_MAPPING[localStorageLanguage]
    : ENTranslations,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'TABLE_EXPAND':
      return { ...state, tableExpanded: true };
    case 'TABLE_COLLAPSE':
      return { ...state, tableExpanded: false };
    case 'SET_SNACKBAR':
      return { ...state, snackbar: { value: action.value, text: action.text } };
    case 'HIDE_HEADER':
      return { ...state, hideHeader: true };
    case 'SHOW_HEADER':
      return { ...state, hideHeader: false };
    case 'SET_LANGUAGE':
      return {
        ...state,
        language: action.value,
        translations: constants.LANGUAGE_MAPPING[action.value],
      };
    default:
      return { ...state };
  }
}

export { initialState };
