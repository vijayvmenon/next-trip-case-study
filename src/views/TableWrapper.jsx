import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import TableHeader from '../components/TableHeader';
import Table from '../components/Table';
import constants from '../constants';

const useStyles = createUseStyles({
  tablePadding: {
    padding: '32px 60px',
  },
});
function TableWrapper({
  tableData = '',
  rowField,
  tableHeaderField,
  interval,
  getTableData,
}) {
  const classes = useStyles();
  const { TABLE_COLUMNS: columns } = constants;
  const rows = tableData?.[rowField] ?? null;
  const { description: headerText, stop_id: stopId } =
    tableData?.[tableHeaderField]?.[0] ?? {};
  const refreshTableEveryMinute = useRef(null);

  useEffect(() => {
    // make the table more extensible and usable without interval setup, if interval props are not passed
    if (interval !== null) {
      console.log('useEffect running', tableData, interval);
      /* set the interval after tableData is set from parent component - this avoids setting interval on initial load 
      when tableData is null and also avoids setting subsequent intervals when tableData changes for the same search criteria,
      as this would keep on adding more interval's which is not required. We ony need the initial interval that was set .
      This also prevents setting Interval when the API returns error */
      if (
        tableData !== null &&
        // interval !== 'error' &&
        interval === 'on-search'
        // && interval !== 'from-interval' &&
        // interval !== 'dropdown-changed'
      ) {
        /* We need to call setInrvl in Parent before the Interval starts so that if the user updates search parameters
         before the first interval starts, the interval is cleared in the useEffect below */
        console.log('before setting intrval', tableData, interval);
        refreshTableEveryMinute.current = setInterval(() => {
          console.log('interval run', tableData, interval);
          getTableData('from-interval');
        }, constants.INTERVAL_FREQUENCY);
      }
      /* clear interval when a new value is searched . This avoids clearing interval on initial load as well and also clears interval 
      when there is API error or when the dropdown is changed */
      if (
        tableData === null &&
        (interval === 'on-search' ||
          interval === 'dropdown-changed' ||
          interval === 'error')
      ) {
        if (refreshTableEveryMinute?.current) {
          console.log('clearing Interval', tableData, interval);
          clearInterval(refreshTableEveryMinute.current);
        }
        refreshTableEveryMinute.current = null;
      }
    }
    return undefined; // for eslint consistent-return rule fix
  }, [tableData, interval]);

  // clearing the interval on component unmount as well
  useEffect(
    () => () => {
      console.log('clearing Interval on unmount');
      clearInterval(refreshTableEveryMinute.current);
      refreshTableEveryMinute.current = null;
    },
    [],
  );
  return (
    <div className={classes.tablePadding}>
      {rows && (
        <>
          {tableHeaderField !== null && (
            <TableHeader headerText={headerText} stopId={stopId} />
          )}
          <Table rows={rows} columns={columns} />
        </>
      )}
    </div>
  );
}

TableWrapper.defaultProps = {
  tableData: '',
  tableHeaderField: null,
  interval: null,
  getTableData: () => {},
};

TableWrapper.propTypes = {
  tableData: PropTypes.oneOfType([
    PropTypes.objectOf(PropTypes.any),
    PropTypes.string,
  ]),
  rowField: PropTypes.string.isRequired,
  tableHeaderField: PropTypes.string,
  interval: PropTypes.string,
  getTableData: PropTypes.func,
};

export default React.memo(TableWrapper);
