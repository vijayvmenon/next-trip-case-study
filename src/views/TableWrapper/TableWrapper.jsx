import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { createUseStyles } from 'react-jss';
import TableHeader from '../../components/Table/HeaderSection';
import Table from '../../components/Table/Table';
import constants from '../../constants';
import Loader from '../../components/UI/Loader';

const useStyles = createUseStyles({
  tablePadding: {
    padding: '32px 8px',
    position: 'relative',
    '@media(min-width: 768px)': {
      padding: '32px 60px',
    },
  },
});
function TableWrapper({
  tableData = '',
  rowField,
  tableHeaderField,
  interval,
  getTableData,
  loader,
}) {
  const classes = useStyles();
  const { TABLE_COLUMNS: columns } = constants;
  const rows = tableData?.[rowField] ?? null;
  const { description: headerText, stop_id: stopId } =
    tableData?.[tableHeaderField]?.[0] ?? {};
  const refreshTableEveryMinute = useRef(null);

  useEffect(() => {
    // make the table more extensible and usable without interval setup, if interval props are not passed
    if (interval !== null) {
      /* set the interval after tableData is set from parent component - this avoids setting interval on initial load 
      when tableData is null and also avoids setting subsequent intervals when tableData changes for the same search criteria,
      as this would keep on adding more interval's which is not required. We ony need the initial interval that was set .
      This also prevents setting Interval when the API returns error */
      if (tableData !== null && interval === 'on-search') {
        /* We need to call setInrvl in Parent before the Interval starts so that if the user updates search parameters
         before the first interval starts, the interval is cleared in the useEffect below */
        refreshTableEveryMinute.current = setInterval(() => {
          getTableData('from-interval');
        }, constants.INTERVAL_FREQUENCY);
      }
      /* clear interval when a new value is searched . This avoids clearing interval on initial load as well and also clears interval 
      when there is API error or when the dropdown is changed */
      if (
        tableData === null &&
        (interval === 'on-search' ||
          interval === 'dropdown-changed' ||
          interval === 'error' ||
          interval === 'reset')
      ) {
        if (refreshTableEveryMinute?.current) {
          clearInterval(refreshTableEveryMinute.current);
        }
        refreshTableEveryMinute.current = null;
      }
    }
    return undefined; // for eslint consistent-return rule fix
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tableData, interval]);

  // clearing the interval on component unmount as well
  useEffect(
    () => () => {
      clearInterval(refreshTableEveryMinute.current);
      refreshTableEveryMinute.current = null;
    },
    [],
  );

  return (
    <>
      {loader && <Loader top="20%" />}
      {rows && (
        <div className={classes.tablePadding}>
          {tableHeaderField !== null && (
            <TableHeader headerText={headerText} stopId={stopId} />
          )}
          <Table rows={rows} columns={columns} />
        </div>
      )}
    </>
  );
}

TableWrapper.defaultProps = {
  tableData: '',
  tableHeaderField: null,
  interval: null,
  getTableData: () => {},
  loader: false,
};

TableWrapper.propTypes = {
  tableData: PropTypes.oneOfType([
    PropTypes.objectOf(PropTypes.any),
    PropTypes.string,
  ]),
  rowField: PropTypes.string.isRequired,
  tableHeaderField: PropTypes.string,
  interval: PropTypes.string,
  getTableData: PropTypes.func,
  loader: PropTypes.bool,
};

export default React.memo(TableWrapper);
