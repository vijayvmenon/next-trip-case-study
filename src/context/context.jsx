import React, { useReducer, createContext } from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';

export const Context = createContext();

export default function AppContext({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
}

AppContext.propTypes = {
  children: PropTypes.objectOf(PropTypes.any).isRequired,
};
