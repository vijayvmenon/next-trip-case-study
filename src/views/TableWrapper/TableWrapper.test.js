import { render, fireEvent, screen } from '@testing-library/react';
import TableWrapper from './TableWrapper';
import renderer from 'react-test-renderer';
import Rows from '../../mockData/rows.json';
import { Context } from '../../context';

const dispatch = jest.fn(); // mock dispatch

const props = {
  tableData: Rows,
  rowField: 'departures',
  tableHeaderField: 'test',
  interval: '',
  getTableData: jest.fn(),
  loader: false,
};

describe('Component:TableWrapper.jsx', () => {
  test('TableWrapper.jsx renders correctly', () => {
    const tableWrapper = renderer
      .create(
        <Context.Provider value={{ dispatch }}>
          <TableWrapper {...props} />
        </Context.Provider>,
      )
      .toJSON();
    expect(tableWrapper).toMatchSnapshot();
  });
});
