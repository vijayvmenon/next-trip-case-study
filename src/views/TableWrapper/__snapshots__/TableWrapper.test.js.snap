// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:TableWrapper.jsx TableWrapper.jsx renders correctly 1`] = `
<div
  className="tablePadding-0-2-1"
>
  <div
    className="root-0-2-2"
  >
    <h3
      className="headerText-0-2-4"
    >
      
    </h3>
    <div
      className="stopId-0-2-3"
    >
      <div
        className="stopIdText-0-2-5"
      />
      <p
        className="stopText-0-2-6"
      >
        0
      </p>
    </div>
  </div>
  <div
    className="tableContainer-0-2-7"
  >
    <table
      className="root-0-2-8"
    >
      <tbody>
        <tr
          className="tableRow-0-2-14"
        >
          <td
            className="tableTd-0-2-23"
          >
            <div
              className="stackedRoot-0-2-20"
            >
              <label
                className="label-0-2-21"
                htmlFor="route_short_name"
              />
              <p
                className="tableCellText-0-2-22"
                id="route_short_name"
              >
                Blue
              </p>
            </div>
            <div
              className="stackedRoot-0-2-20"
            >
              <label
                className="label-0-2-21"
                htmlFor="description"
              />
              <p
                className="tableCellText-0-2-22"
                id="description"
              >
                to Mpls-Target Field
              </p>
            </div>
            <div
              className="stackedRoot-0-2-20"
            >
              <label
                className="label-0-2-21"
                htmlFor="departure_text"
              />
              <p
                className="tableCellText-0-2-22"
                id="departure_text"
              >
                <div
                  className="root-0-2-24"
                >
                  <svg
                    className="blinker-0-2-25"
                    data-testid="broadcast-svg"
                  >
                    broadcast.svg
                  </svg>
                  <div>
                    Due
                  </div>
                </div>
              </p>
            </div>
          </td>
        </tr>
        <tr
          className="tableRow-0-2-14"
        >
          <td
            className="tableTd-0-2-23"
          >
            <div
              className="stackedRoot-0-2-20"
            >
              <label
                className="label-0-2-21"
                htmlFor="route_short_name"
              />
              <p
                className="tableCellText-0-2-22"
                id="route_short_name"
              >
                Blue
              </p>
            </div>
            <div
              className="stackedRoot-0-2-20"
            >
              <label
                className="label-0-2-21"
                htmlFor="description"
              />
              <p
                className="tableCellText-0-2-22"
                id="description"
              >
                to Mall of America
              </p>
            </div>
            <div
              className="stackedRoot-0-2-20"
            >
              <label
                className="label-0-2-21"
                htmlFor="departure_text"
              />
              <p
                className="tableCellText-0-2-22"
                id="departure_text"
              >
                <div
                  className="root-0-2-24"
                >
                  <svg
                    className="blinker-0-2-25"
                    data-testid="broadcast-svg"
                  >
                    broadcast.svg
                  </svg>
                  <div>
                    7 Min
                  </div>
                </div>
              </p>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
